src/main.c

```diff
@@ -1,30 +1,145 @@
-#include <stdlib.h>		// malloc
-#include <string.h>		// strlen
-#include <stdbool.h>
-
 #include <arpa/inet.h>
 #include <netinet/in.h>
-#include <netinet/ip.h>				// provides declarations for ip header
-#include <netinet/if_ether.h>		// for ETH_P_ALL
-#include <net/ethernet.h>			// for ether_header
+#include <netinet/ip.h>
+#include <netinet/if_ether.h>
+#include <net/ethernet.h>
 #include <net/if.h>
 #include <sys/socket.h>
 #include <sys/ioctl.h>
 #include <sys/time.h>
 #include <sys/types.h>
-#include <unistd.h>
 #include <signal.h>
 #include <byteswap.h>
 
+#include "ns_config.h"
 #include "ns_arp.h"
 #include "ns_arp_packet.h"
-
 #include "array.h"
 #include "functions.h"
 
+#define DEBUG_PRINT(...) do { if (m.debug) printf(__VA_ARGS__); } while(0)
+
+// Main program state
+struct main {
+    char *eth_dev_s;
+    char *eth_ip_s;
+    char *broadcast_ip_s;
+    char *subnet_s;
+    char *allow_gateway_s;
+
+    struct target *target_list;
+
+    uint32_t *source_blacklist;
+    uint32_t *source_allowlist;
+
+    // DNS cache for hostnames in allow/exclude lists
+    dns_cache_t exclude_dns_cache;
+    dns_cache_t include_dns_cache;
+    int dns_cache_refresh_interval;
+
+    unsigned char eth_ip[4];
+    unsigned char gate_ip[4];
+
+    unsigned int  subnet;
+
+    unsigned char *buffer;
+    int sock_raw;
+    bool alive;
+    bool debug;
+} m;
+// Utility: resolve hostname to IPv4 address (returns 0 on success)
+int resolve_hostname(const char *hostname, uint32_t *out_ip) {
+	if (!hostname || !out_ip) return -1;
+	
+	struct addrinfo hints, *res = NULL;
+	memset(&hints, 0, sizeof(hints));
+	hints.ai_family = AF_INET;
+	hints.ai_socktype = SOCK_STREAM;
+	
+	int err = getaddrinfo(hostname, NULL, &hints, &res);
+	if (err != 0 || !res) return -1;
+	
+	struct sockaddr_in *addr = (struct sockaddr_in *)res->ai_addr;
+	if (!addr || res->ai_addrlen < sizeof(struct sockaddr_in)) {
+		freeaddrinfo(res);
+		return -1;
+	}
+	
+	*out_ip = addr->sin_addr.s_addr;
+	freeaddrinfo(res);
+	return 0;
+}
+
+// Non-blocking DNS cache refresh (only refresh if really needed and been a while)
+void refresh_dns_cache_if_needed(dns_cache_t *cache) {
+	if (!cache) return;
+	
+	time_t now = time(NULL);
+	if (cache->last_refresh == 0 || (now - cache->last_refresh) >= DEFAULT_DNS_CACHE_REFRESH_INTERVAL) {
+		DEBUG_PRINT("Refreshing DNS cache with %zu entries\n", cache->count);
+		
+		// Bounds check to prevent buffer overflow
+		size_t safe_count = (cache->count > MAX_SOURCE_LIST_ENTRIES) ? MAX_SOURCE_LIST_ENTRIES : cache->count;
+		
+		for (size_t i = 0; i < safe_count; ++i) {
+			// Additional safety checks
+			if (!cache->entries[i].entry || strlen(cache->entries[i].entry) == 0) continue;
+			
+			if (cache->entries[i].is_hostname) {
+				DEBUG_PRINT("Resolving hostname: %s\n", cache->entries[i].entry);
+				uint32_t resolved_ip;
+				if (resolve_hostname(cache->entries[i].entry, &resolved_ip) != 0) {
+					cache->entries[i].ip = 0; // Mark as unresolved
```

Comment:
Not strictly a problem, but I would like if this weren't a blind magic value.

Reply:
Plus wouldn't this assumption fail if we resolved 0.0.0.0 somehow?

Reply:
Now that I'm thinking about it, that would never make sense anyway since there's no need to wake up localhost.